Posts/2021/04/NewTemporaryFoldersPowerShell

```diff
@@ -0,0 +1,89 @@
+
+However, we can “Borrow” the filename created and use it to create a folder instead.   It’s not really difficult, but maybe just not thought of very often.
+
+`New-TemporaryFile`
+
+It will create a file like below.
+
+Mode                 LastWriteTime         Length Name
+----                 -------------         ------ ----
+-a----         3/31/2021   9:25 PM              0 tmpA927.tmp
+
+ 
+Ok, that really wasn’t that impressive but what if we were to do this instead?
+
+`$File=New-TemporaryFile`
+
+Now we’ve created the file and stored it away in the $File object.   With this we can remove the File of course using the Remove-Item Cmdlet
+
+`Remove-Item -path $File -force`
+
+AHA!  I’ve all of that hard work.  But the $File object is still there to use.  So I could access the name and use it to create a directory instead in the following manner
+New-Item -itemtype Directory -path $File.Name
+
+But the problem is that it would be in whatever default folder PowerShell was looking into at the time.
+Hmmmmm…. How to solve that?
+But there is a built in variable call $ENV:Temp which targets the exact Temporary folder that the New-TemporaryFile uses too!
+I can then take that variable and the original name of the Temporary file and combine them together like this
+
+$ENV:Temp + ‘\’ + $File.Name
+
+*or*
+
+I can even put them together in a single String like this
+
+`"$($ENV:Temp)\$($File.Name)"`
+
+With this I could just create a New temporary Directory under our temp folder like this
+
+`New-Item -itemtype Directory -path "$($ENV:Temp)\$($File.Name)"`
+
+Now to track it, I could same thing as last time if I wanted to know the "Random Directory name" to later remove it
+
+`$Folder=New-Item -itemtype Directory -path "$($ENV:Temp)\$($File.Name)"`
+
+Then when I am done with that folder that was presumably used to hold some garbage data, I can just use 
+Remove-Item again. 
+
+But because it's a directory, I need to add -recurse -force to ensure all data and Subfolders are removed.
+
+`Remove-Item -path $Folder -recurse -force`
+
+But here is the fun and neat bit.  If you needed on a regular basis, we can make this into a quick function for your code, module or to impress friends with!
+
+```
+Function New-TemporaryFolder {
+    # Create Temporary File and store object in $t
+    $t = New-TemporaryFile
+
+    # Remove the Temporary File .... Muah ha ha ha haaaaa!
+    Remove-item $t -force
+
+    # Make a new folder based upon the old name
+    New-Item -itemtype Directory -path "$($ENV:Temp)\$($File.Name)"
+}
+```
+
+Why did all of this pop into my head?  I was actually creating some PowerShell for customer and needed a consistent and random set of folders in a common and easily erasable location. 
+
+I was hoping that the we had a "New-TemporaryDirectory" Cmdlet, but found it was just as easy to write one by "Borrowing" and existing Cmdlet :)
```

Comment:
I was hoping that we had a `New-TemporaryDirectory` cmdlet, but found it was just as easy to write one by "borrowing" an existing Cmdlet :)